### PR TITLE
[Host ir] support for set reduce and binary op

### PR DESCRIPTION
**Pull request by samnordmann**
This PR belongs to a series of stacked PRs:
1. #4144
2. #4145
3. **=> You are here:** #4146
4. #4147

Add support for `LoadStoreOp`, `BinaryOp`, `ReductionOp`, including support for pre-allocated output, which is not provided by ExprEvaluator.

### CODE REVIEW

**Comment by samnordmann** [comment]
!test

File: csrc/host_ir/executor.cpp [hunk 1]
```diff
@@ -570,6 +570,31 @@ void HostIrEvaluator::handle(LinearOp* linear) {
...
+
+  // If output has root domain, compute and apply permutation
+  if (out_tv->hasRoot()) {
+    auto permutation = ir_utils::computePermutation(
+        out_tv->getRootDomain(), out_tv->getLogicalDomain());
+    NVF_ERROR(
+        permutation.has_value(),
+        "The logical domain of a Set.Permute is supposed to be a permutation of the root domain: ",
+        out_tv->toString());
+    in_tensor = in_tensor.permute(*permutation).contiguous();
```
Comment (samnordmann) [inline csrc/host_ir/executor.cpp:587]: note that the `.contiguous()` is necessary here, and I think this is an unexposed bug in `LoadStoreOp::evaluate()` -- however fixing it there incidentally causes another test failure.

The bug was not exposed because we never "host evaluate" a set.Permute op before this PR

File: tests/cpp/test_multidevice_pipeline.cpp [hunk 2]
```diff
@@ -457,135 +457,4 @@ INSTANTIATE_TEST_SUITE_P(
         testing::Values(0, 1),
         testing::Values(true)));
 
-// Different scheduling modes used in
```
Comment (samnordmann) [inline tests/cpp/test_multidevice_pipeline.cpp:460]: This test is not relevant anymore since we don't use generated kernels for now. So we'll add it back in times if we think this is useful. But in the meantime it is just technical debt

File: csrc/host_ir/executor.cpp (hunk 1 above)
Comment (wujingyue) [reply csrc/host_ir/executor.cpp:587]: What was the symptom and can you write a standalone repro so I can help here? 

> The bug was not exposed because we never "host evaluate" a set.Permute op before this PR

I'm sure many tests in https://github.com/NVIDIA/Fuser/blob/4b3dad22d32ce3c338db2e17fd07cec68467cf4b/tests/cpp/test_alias.cpp#L4 "host evaluate" Set.Permute (aka transpose). 

I suspect the bug is not with LoadStoreOp::evaluate() but with some downstream code assuming the input tensor being contiguous. Therefore, a standalone repro would help a lot.

Comment (samnordmann) [reply csrc/host_ir/executor.cpp:587]: Nevermind, the problem actually comes from that `at::permute` is an alias operation, but the operation must be realized before a communication, since nccl/ucc only look at the pointer.

IIUC, before this patch, a permute op would generate a kernel, which does not produce an alias. Now that it's going through Aten, we get an alias. 

If we remove `.contiguous()` in the host Ir evaluator, then, for example, test `Scatter/PipelineTestTwoStages.Communication/10`
```
%HostIrContainer { (T0_g_float[iS0{3}, iS1{4}, iS2{3}, iS3{5}] (DeviceMesh{0})) -> (T3_g_float[iS11{3}, ideviceIdx.x12{4}, iS13{3}] (DeviceMesh{0 1 2 3})) :
  T1_l_float[iS4{3}, iS5{4}, iS6{3}, rS7{5}] (DeviceMesh{0})
     = reduction( T0_g_float[iS0{3}, iS1{4}, iS2{3}, iS3{5}] (DeviceMesh{0}), op = add, initial value = float(0), allreduce = false )
  T4_g_float[iS15{4}, iS14{3}, iS16{3}] (DeviceMesh{0})
     = Set.Permute( T1_l_float[iS4{3}, iS5{4}, iS6{3}, rS7{5}] (DeviceMesh{0}), cache_op=Streaming )
  T5_g_float[ideviceIdx.x17{4}, iS18{3}, iS19{3}] (DeviceMesh{0 1 2 3}) = ALLOCATE(buffer=T5_g_float[ideviceIdx.x17{4}, iS18{3}, iS19{3}] (DeviceMesh{0 1 2 3}), mem_type=global, size=36, zero_init=false, resets_to_zero=false)
  Communication 23 (type=Scatter, team=(0 1 2 3), root=0, input=T4_g_float[iS15{4}, iS14{3}, iS16{3}] (DeviceMesh{0}), output=T5_g_float[ideviceIdx.x17{4}, iS18{3}, iS19{3}] (DeviceMesh{0 1 2 3}), backend=NCCL)
  Wait Communication 23
  T6_l_float[iS21{3}, ideviceIdx.x20{4}, iS22{3}] (DeviceMesh{0 1 2 3})
     = Set.Permute( T5_g_float[ideviceIdx.x17{4}, iS18{3}, iS19{3}] (DeviceMesh{0 1 2 3}), cache_op=Streaming )
  T3_g_float[iS11{3}, ideviceIdx.x12{4}, iS13{3}] (DeviceMesh{0 1 2 3})
     = T6_l_float[iS21{3}, ideviceIdx.x20{4}, iS22{3}] (DeviceMesh{0 1 2 3})
     + T6_l_float[iS21{3}, ideviceIdx.x20{4}, iS22{3}] (DeviceMesh{0 1 2 3});
} // %HostIrContainer
```

 will fail because the input tensor at node 0 is not contiguous.

I can either leave the `.contiguous()` as is to force the op to be realized, or add it in `postScatter` etc. I'm fine either way (those cases are really corner cases that we do not especially target for now), let me know

**Comment by samnordmann** [comment]
!test

File: csrc/host_ir/executor.cpp (hunk 1 above)
Comment (samnordmann) [reply csrc/host_ir/executor.cpp]: I tried to move `contiguous()` in `postScatter` in this commit [move .contiguous to be in postScatter](https://github.com/NVIDIA/Fuser/pull/4146/commits/85b7b751b82c31f5b019773d5adfdec21210ab8b), but I'm reverting it because:
1. The same fix needs to be done in `postReduceScatter` as well, see [CI failures](https://gitlab-master.nvidia.com/dl/pytorch/fuser-gh-mirror/-/jobs/158579157/raw) 
2. Since a recent PR https://github.com/NVIDIA/Fuser/pull/4170, the error is now intercepted earlier at https://github.com/NVIDIA/Fuser/blob/f44f1913c26f8325099ab6fe46d678cbea435658/csrc/host_ir/executor.cpp#L456

Therefore I revert the commit and keep `.contiguous()` in `HostIrEvaluator::handle(LoadStoreOp*)`. We might want to refine this behavior in the future (i.e. control whether to realise or alias the permute), but it seems like the safe option for now.

(Alternatively, a fix could be to add `.contiguous()` on the input here https://github.com/NVIDIA/Fuser/blob/f44f1913c26f8325099ab6fe46d678cbea435658/csrc/host_ir/executor.cpp#L447.)

**Comment by samnordmann** [comment]
!build

**Comment by samnordmann** [comment]
!test

File: csrc/host_ir/executor.cpp (hunk 1 above)
Comment (wujingyue) [reply csrc/host_ir/executor.cpp:648]: > IIUC, before this patch, a permute op would generate a kernel, which does not produce an alias.

That sounds like the right behavior. AFAIU, reorder-sharded-axis and make-resharding-contiguous kicked in. When seeing the allocation domain not in favor, nvFuser considers the permute op as a non-meta op and therefore generates a kernel. 

> Now that it's going through Aten, we get an alias.

That sounds like a bug somewhere. Assuming reorder-sharded-axis and make-resharding-contiguous still kicked in as before, why didn't nvFuser generate a kernel for the Set.Permute?

File: csrc/host_ir/executor.cpp [hunk 3]
```diff
@@ -654,6 +679,80 @@ void HostIrEvaluator::handle(kir::Allocate* allocate) {
...
+      at::sub_out(output, lhs, rhs);
+      break;
+    case BinaryOpType::Mul:
+      at::mul_out(output, lhs, rhs);
+      break;
+    case BinaryOpType::Div:
+      at::div_out(output, lhs, rhs);
+      break;
+    default:
+      NVF_CHECK(
```
Comment (wujingyue) [inline csrc/host_ir/executor.cpp:706]: NVF_THROW

File: csrc/host_ir/executor.cpp [hunk 4]
```diff
@@ -654,6 +679,80 @@ void HostIrEvaluator::handle(kir::Allocate* allocate) {
...
+      at::sum_out(output, input, reduction_axes);
+      return;
+    case BinaryOpType::Max:
+      at::amax_out(output, input, reduction_axes);
+      return;
+    case BinaryOpType::Min:
+      at::amin_out(output, input, reduction_axes);
+      return;
+    default:
+      NVF_CHECK(
```
Comment (wujingyue) [inline csrc/host_ir/executor.cpp:747]: ditto

File: csrc/host_ir/executor.cpp (hunk 1 above)
Comment (samnordmann) [reply csrc/host_ir/executor.cpp:648]: > That sounds like the right behavior. AFAIU, reorder-sharded-axis and make-resharding-contiguous kicked in. When seeing the allocation domain not in favor, nvFuser considers the permute op as a non-meta op and therefore generates a kernel.

Yes

> >Now that it's going through Aten, we get an alias.

> That sounds like a bug somewhere. Assuming reorder-sharded-axis and make-resharding-contiguous still kicked in as before, why didn't nvFuser generate a kernel for the Set.Permute?

It's just because with the current patch, a `LoadStoreOp` now goes through a dedicated `HostIrEvaluator::handle` and not a `PostOnStream(HostUnit(.))`, where ATen is used and not an nvfuser generated kernel.

IMO this is equally good, as long as we make sure that the op is not an alias op when it shouldn't be.

Comment (wujingyue) [reply csrc/host_ir/executor.cpp:648]: > IMO this is equally good

Letting LoadStoreOp go through HostIrEvaluator::handle means it'll be interpreted and we'll lose the opportunity to fuse it into an upstream op as one kernel. So it's not equally good. However, I don't mind suboptimal performance as a shortcut. I guess you had to avoid PostOnStream because nvFuser can't generate a kernel with stream parallel type yet. 

A bigger issue is that `.contiguous` is applied unconditionally for all transposes. I believe this will cause wrong numbers because allocation doesn't necessarily equal logical. I think this can be fixed by 
```
if result_of_permute doesn't match allocation domain:
  create the output tensor according to the allocation domain
  out_tensor.copy_(result_of_permute)
```

What should happen eventually is the following: 
1. As has been done by https://github.com/NVIDIA/Fuser/blob/857d1df833b70b38ca31e3640016b436c972503b/csrc/scheduler/expr_eval_sched.cpp#L43, segmentation assigns Set.Permute to SchedulerType::ExprEval when its input and output allocation domain support it to be an alias op (cf. ). 
2. As has been done by https://github.com/NVIDIA/Fuser/blob/857d1df833b70b38ca31e3640016b436c972503b/csrc/runtime/fusion_kernel_runtime.cpp#L545-L546, host IR lowering copies an aliasing Set.Permute to top-level and compiles a non-aliasing Set.Permute into a kernel (together with surrounding ops that belong to the same segment) 
3. As a missing piece, the compilation (scheduling and codegen) has to support stream parallel type.

Comment (wujingyue) [reply csrc/host_ir/executor.cpp:648]: > A bigger issue

To unblock your experiments, I'd consider subclassing HostIrEvaluator so this issue doesn't affect FusionExecutorCache.

Comment (samnordmann) [reply csrc/host_ir/executor.cpp:648]: > I guess you had to avoid PostOnStream because nvFuser can't generate a kernel with stream parallel type yet.

Right, and also because PostOnStream's either
1. either goes through FusionExecutorCache, and then its support for pre-allocated output is through an extra unnecessary copy (leaving perf on the table).
2. or goes through FusionExecutor with default scheduling and the GPU kernel is not aliasing and is fully sequential (therefore extremely slow)

In both cases we lose performance, that is why I don't think  generating a kernel with `PostOnStream(HostUnit(.))` is necessarily better than using ATen.

This being said, sorry for my ignorance, but I feel we are speculating here because I don't know any concrete target  involving these considerations, in which this or that solution is better. This is what I meant by "equally good" earlier, but I'm sorry it wasn't clear at all. Do you have an example in mind?

Supporting "Set" in host IR as in this patch follows the logic of incrementally adding the support for more ops, as I did previously for many ops in HostIr. I wasn't expecting that `SetOp` would raise more concerns than other ops, nor that we had any short term plan in MultiGPU roadmap to use generated kernels. Supporting "Set" thorugh ATen doesn't map to a precise target on our roadmap either ; I just use `set` in the tests to exercise stream lowering and other logics, and I chose to use `set` just because it is an easy unary op with simple semantics. If this raises too much concerns, I can always use another unary op (e.g. `abs`).

Please let me know what I need to do so we can merge this PR quickly and move on to upstreaming the stream lowering logic.


> A bigger issue is that .contiguous is applied unconditionally for all transposes. I believe this will cause wrong numbers because allocation doesn't necessarily equal logical. I think this can be fixed by

The PR passes all the tests. Could you write a test triggering this bug? Can it be done in a future PR?

> if result_of_permute doesn't match allocation domain:
>   create the output tensor according to the allocation domain
>   out_tensor.copy_(result_of_permute)


I am not sure to understand what I need to check exactly. Should I check if Root = allocation domain? Can you explain with more details?
I am also ok to simply "assert" we are not in a problematic case


> To unblock your experiments, I'd consider subclassing HostIrEvaluator so this issue doesn't affect FusionExecutorCache.

Do you mean using a different HIR lowering path in `MultiDeviceExecutor` and in `FusionExecutorCache`? Bottom line, this would mean using different `isLowerableAsStandaloneHostOp` in the two instances, and in all the passes. I am ok to do if this is the way to unblock me -- this is already the purpose of `MultiDeviceExecutor` to enable development without raising too much concerns on FusionExecutorCache. But I am not sure to understand the motivation, so I don't want to spend too much time doing that. The present patch is not breaking any test or causing any performance regression. It doesn't even change the default behavior of `FusionExecutorCache`, where HIR lowering is off by default.

Let me know what you want me to do.

Comment (samnordmann) [reply csrc/host_ir/executor.cpp:648]: Another solution that would be fine with me is to only lower as a standalone host op the `LoadStoreOp`s that do not contain any permute

Comment (wujingyue) [reply csrc/host_ir/executor.cpp:648]: > Could you write a test triggering this bug? 

Sure, I'll try that tomorrow. I suspect some of the tests in test_alias.cpp would fail if host IR lowering was on.

**Comment by samnordmann** [comment]
!test

File: csrc/host_ir/executor.cpp (hunk 1 above)
Comment (wujingyue) [reply csrc/host_ir/executor.cpp]: > Sure, I'll try that tomorrow. I suspect some of the tests in test_alias.cpp would fail if host IR lowering was on.

I managed to reproduce the error based on this branch. 

Patch
```
diff --git a/tests/cpp/test_alias.cpp b/tests/cpp/test_alias.cpp
index 04f034f1..559dbafd 100644
--- a/tests/cpp/test_alias.cpp
+++ b/tests/cpp/test_alias.cpp
@@ -150,6 +150,9 @@ TEST_F(AliasTest, View_NoAliasForIncompliantLayout) {
 }

 TEST_F(AliasTest, ViewPermute) {
+  EnableOptionsGuard opt_guard;
+  EnableOptionsGuard::getCurOptions().set(EnableOption::HostIrLowering);
+
   auto fusion = std::make_unique<Fusion>();
   FusionGuard fg(fusion.get());
```
and then run
```
$ NVFUSER_DUMP=host_ir bin/test_nvfuser --gtest_filter=AliasTest.ViewPermute

Note: Google Test filter = AliasTest.ViewPermute
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from AliasTest
[ RUN      ] AliasTest.ViewPermute

%HostIrContainer { (T0_g_float[iS0{2}, iS1{3}, iS2{4}]) -> (T2_g_float[iS10{12}, iS9{2}]) :
  T1_l_float[iS3{2}, iS8{12}rf] = view( T0_g_float[iS0{2}, iS1{3}, iS2{4}] )
  T2_g_float[iS10{12}, iS9{2}]
     = Set.Permute( T1_l_float[iS3{2}, iS8{12}rf], cache_op=Streaming )
} // %HostIrContainer

Aliases:{
}
unknown file: Failure
C++ exception with description " INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/host_ir/executor.cpp":645, please report a bug with repro script to NVFuser at https://github.com/NVIDIA/Fuser/issues. the set opT2_g_float[iS10{12}, iS9{2}]
   = Set.Permute( T1_l_float[iS3{2}, iS8{12}rf], cache_op=Streaming )
must not be a permute
```

If the same patch is based on main, the test passes.

Comment (wujingyue) [reply csrc/host_ir/executor.cpp]: I pushed a change which hopefully resolve this. PTAL

**Review by wujingyue — APPROVED** [review]
LGTM otherwise

File: csrc/host_ir/executor.cpp (hunk 1 above)
Comment (samnordmann) [reply csrc/host_ir/executor.cpp]: LGTM, thank you very much!

**Comment by samnordmann** [comment]
!test
